### ISSUE TITLE
CodeAction erroneously parsed as Command

### ISSUE THREAD

**Issue by mickaelistria** [eclipse/lsp4j#219]
I'm trying to adopt the new CodeAction class in LSP4E.

In my tests, I set the codeActions like this in my LS
```java
TextEdit tEdit = new TextEdit(new Range(new Position(0, 0), new Position(0, 5)), "fixed");
WorkspaceEdit wEdit = new WorkspaceEdit(Collections.singletonMap(f.getLocationURI().toString(), Collections.singletonList(tEdit)));
MockLanguageSever.INSTANCE.setCodeActions(Collections.singletonList(Either.forRight(new CodeAction("fixme", "fix", Collections.emptyList(), wEdit, null))));
```
I can see when debugging my LS that this object is still the one returned on server-side upon `codeActions(...)` request.

In the client, when I invoke `codeActions(...)`, the returned object seems to be erroneous in some cases, or generate a parsing error in some others. For instance, here is the response I get with the codeAction described about
```json
{
  "jsonrpc": "2.0",
  "id": "3",
  "result": [
    {
      "title": "fixme"
    }
  ]
}
```
As you can see, it's missing a lot of content. With more logging, I can see the following message is tranferred from server to client
```json
{"jsonrpc":"2.0","id":"3","result":[{"title":"fixme","kind":"fix","diagnostics":[],"edit":{"changes":{"file:/home/mistria/junit-workspace/CodeActionTests1533196529057/test1533196529126.lspt":[{"range":{"start":{"line":0,"character":0},"end":{"line":0,"character":5}},"newText":"fixed"}]}}}]}
```
which seems correct.

It looks like since `title` is a common node to both `Command` and `CodeAction`, the parser decides to pick Command instead of CodeAction, without looking at further nodes that can be used to discriminate what's the right type.

**Comment by mickaelistria** [comment]
With some other cases, like `edit` being null and `command` being set in the CodeAction, this even turns into a heavy ParseError.

**Comment by spoenemann** [comment]
The problem is the same as with #226: The LSP has been developed in a way that makes it very hard to parse the data structures with a statically typed language such as Java. Our current implementation of `Either` doesn't support differentiating between two object types.

**Comment by mickaelistria** [comment]
Could parsing a Either start to initiate both objects on parsing and reject the one that doesn't match?
